### PR TITLE
Infra 3707 es statuscheck

### PR DESCRIPTION
**Pull request by elainearbaugh**

### CODE REVIEW

**Comment by elainearbaugh** [comment]
trigger build

**Comment by elainearbaugh** [comment]
@mahendra this is ready for review now (and up on stage!) It's long but there are a lot of test fixtures... There are a few things from my other active PRs in this so those should probably be reviewed first

Right now you can only make ElasticsearchStatusChecks through the admin site (since most people will want to make ESchecks through Grafana), but I could create views for the forms if we want them. 

Includes
- ElasticsearchStatusCheck model
- input query validation (with a specific format for queries specified)
- Get and parse response from ES
- Admin ESStatusCheck form
- Lots of testing

File: cabot/metricsapp/api/elastic.py [hunk 1]
```diff
@@ -1,11 +1,54 @@
 from elasticsearch import Elasticsearch
+from enum import Enum
+
+
+class SupportedMetrics(Enum):
```
Comment (mahendra) [inline cabot/metricsapp/api/elastic.py]: How does this being a subclass of `Enum` help? It will work with `SupportedMetrics(object)` too.

Comment (mahendra) [inline cabot/metricsapp/api/elastic.py]: Would request more doc strings on this class - to understand what this means. What is the difference between single and multiple? How do we decide on one v/s another? - based on `value` v/s `values` alone?

File: cabot/metricsapp/api/elastic.py [hunk 2]
```diff
@@ -1,11 +1,54 @@
...
+    METRICS_SINGLE = ['min', 'max', 'avg', 'value_count', 'sum', 'cardinality', 'moving_avg', 'derivative']
+    # Response includes the field "values"
+    METRICS_MULTIPLE = ['percentiles']
 
 
 def create_es_client(urls, timeout):
     """
     Create an elasticsearch-py client
     :param urls: comma-separated string of urls
+    :param timeout: timeout for queries to the client
```
Comment (mahendra) [inline cabot/metricsapp/api/elastic.py]: should `timeout` be a kwarg with a default?

File: cabot/metricsapp/forms/elastic.py [hunk 3]
```diff
@@ -2,8 +2,10 @@
 from django.forms import ModelForm
 from elasticsearch.client import ClusterClient
 from elasticsearch.exceptions import ConnectionError
-from cabot.metricsapp.api.elastic import create_es_client
-from cabot.metricsapp.models import ElasticsearchSource
+import json
```
Comment (mahendra) [inline cabot/metricsapp/forms/elastic.py]: nit: please move this to the beginning of the file

File: cabot/metricsapp/models/elastic.py [hunk 4]
```diff
@@ -1,6 +1,13 @@
 from django.db import models
-from cabot.metricsapp.api import create_es_client
-from .base import MetricsSourceBase
+from elasticsearch_dsl import Search
+import json
```
Comment (mahendra) [inline cabot/metricsapp/models/elastic.py]: nit: please club all the `import`s together and the `from ... import ...` together..

File: cabot/metricsapp/models/elastic.py [hunk 5]
```diff
@@ -37,3 +49,132 @@ def client(self):
...
+                  - [timestamp, value]
+                check:
+        """
+        parsed_data = dict()
+        parsed_data['raw'] = None
+        # Will be set to true if we encounter an error
+        parsed_data['error'] = False
+        parsed_data['data'] = []
+
+        try:
```
Comment (mahendra) [inline cabot/metricsapp/models/elastic.py]: shouldn't this happen when we try `ElasticsearchStatusCheck().save()` ? and we throw an error that the data is invalid?

Comment (mahendra) [inline cabot/metricsapp/models/elastic.py]: Should we also do a `validate_query` ?

File: cabot/metricsapp/tests/fixtures/es_multiple_metrics_terms.json [hunk 6]
```diff
@@ -0,0 +1,187 @@
+{
+    "hits": {
+        "hits": [
+            {
+                "_score": 11.410242,
+                "_type": "metrics-event",
+                "_id": "a",
+                "_index": "meddlesome-metrics"
```
Comment (mahendra) [inline cabot/metricsapp/tests/fixtures/es_multiple_metrics_terms.json]: I can see the frustration here :-)

File: cabot/metricsapp/tests/fixtures/es_multiple_metrics_terms.json [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+                                "min": {
+                                    "value": 14.243
+                                },
+                                "key_as_string": "2017-04-07T15:00:00.000Z",
+                                "key": 1491577200000,
+                                "doc_count": 600
+                            }
+                        ]
+                    },
+                    "key": "maroon",
```
Comment (mahendra) [inline cabot/metricsapp/tests/fixtures/es_multiple_metrics_terms.json]: (without my glasses) for a moment, I thought you are calling me a `moron`

File: cabot/metricsapp/tests/fixtures/metrics_series.yaml [hunk 8]
```diff
@@ -1,6 +1,4 @@
 error: False
-error_message:
```
Comment (mahendra) [inline cabot/metricsapp/tests/fixtures/metrics_series.yaml]: ?

File: cabot/metricsapp/tests/test_elasticsearch.py [hunk 9]
```diff
@@ -30,3 +38,235 @@ def test_client_whitespace(self):
...
+
+
+def mock_time():
+    return 1491577200.0
+
+
+class TestElasticsearchStatusCheck(TestCase):
+    def setUp(self):
+        self.user = User.objects.create_user('user')
+        self.es_source = ElasticsearchSource.objects.create(
```
Comment (mahendra) [inline cabot/metricsapp/tests/test_elasticsearch.py]: shouldn't we use django test fixtures for this?

File: cabot/metricsapp/forms/elastic.py [hunk 10]
```diff
@@ -22,3 +24,44 @@ def clean_urls(self):
...
+        """
+        queries = self.cleaned_data['queries']
+        try:
+            query_list = json.loads(queries)
+        except ValueError:
+            raise ValidationError('Queries must be json-parsable')
+
+        for query in query_list:
+            try:
+                validate_query(query)
```
Comment (mahendra) [inline cabot/metricsapp/forms/elastic.py]: This should ideally move to the model.save() - In the future, we may support cabot APIs. The validation functionality can be at the core.

File: cabot/metricsapp/forms/elastic.py [hunk 11]
```diff
@@ -22,3 +24,44 @@ def clean_urls(self):
...
+            self.fields['source'].queryset = ElasticsearchSource.objects.all()
+            return ret
+
+    def clean_queries(self):
+        """
+        Make sure input queries are formatted correctly.
+        """
+        queries = self.cleaned_data['queries']
+        try:
+            query_list = json.loads(queries)
```
Comment (mahendra) [inline cabot/metricsapp/forms/elastic.py]: This too can move to model.save()

File: cabot/metricsapp/models/elastic.py [hunk 12]
```diff
@@ -37,3 +49,132 @@ def client(self):
...
+              - series: a.b.c.d
+                datapoints:
+                  - [timestamp, value]
+                  - [timestamp, value]
+              - series: a.b.c.p.q
+                datapoints:
+                  - [timestamp, value]
+                check:
+        """
+        parsed_data = dict()
```
Comment (mahendra) [inline cabot/metricsapp/models/elastic.py]: minor: fastest way to do it
```python
parsed_data = dict(raw=None, error=False, data=[])
```

File: cabot/metricsapp/models/elastic.py [hunk 13]
```diff
@@ -37,3 +49,132 @@ def client(self):
...
+        parsed_data['raw'] = None
+        # Will be set to true if we encounter an error
+        parsed_data['error'] = False
+        parsed_data['data'] = []
+
+        try:
+            queries = json.loads(self.queries)
```
Comment (mahendra) [inline cabot/metricsapp/models/elastic.py]: this is repeated in the form/clean

File: cabot/metricsapp/models/elastic.py [hunk 14]
```diff
@@ -37,3 +49,132 @@ def client(self):
...
+            parsed_data['error_message'] = str(e)
+            parsed_data['error'] = True
+            return parsed_data
+
+        for query in queries:
+            source = ElasticsearchSource.objects.get(name=self.source.name)
+            try:
+                search = Search().from_dict(query)
+                response = search.using(source.client).index(source.index).execute()
+                parsed_data['raw'] = response.to_dict()
```
Comment (mahendra) [inline cabot/metricsapp/models/elastic.py]: response.to_dict() is called twice. We need to avoid it.

```python
raw_data = response.to_dict()
parsed_data['raw'] = raw_data
parsed_data['data'].extend(self._es_rec(raw_data['aggregations']))
```

File: cabot/metricsapp/models/elastic.py [hunk 15]
```diff
@@ -37,3 +49,132 @@ def client(self):
...
+            queries = json.loads(self.queries)
+        except ValueError as e:
+            logger.exception('Error loading Elasticsearch queries: {}'.format(self.queries))
+            parsed_data['error_code'] = type(e).__name__
+            parsed_data['error_message'] = str(e)
+            parsed_data['error'] = True
+            return parsed_data
+
+        for query in queries:
+            source = ElasticsearchSource.objects.get(name=self.source.name)
```
Comment (mahendra) [inline cabot/metricsapp/models/elastic.py]: This can move out of the for loop. We don't need to do a db call in each for loop.

File: cabot/metricsapp/models/elastic.py [hunk 16]
```diff
@@ -37,3 +49,132 @@ def client(self):
...
+            except Exception as e:
+                logger.exception('Error executing Elasticsearch query: {}'.format(query))
+                parsed_data['error_code'] = type(e).__name__
+                parsed_data['error_message'] = str(e)
+                parsed_data['error'] = True
+                break
+
+        return parsed_data
+
+    def _es_rec(self, series, series_name=None):
```
Comment (mahendra) [inline cabot/metricsapp/models/elastic.py]: Would recommend renaming this function. We should mention what the function does. Not the implementation.

In the future, if you change the function to be non-recursive, you will have to change the name.

File: cabot/metricsapp/models/elastic.py [hunk 17]
```diff
@@ -37,3 +49,132 @@ def client(self):
...
+        if len(series) == 0:
+            return []
+
+        # All aggregations should be named 'agg' (validated in validate_query())
+        if series[0].get('agg') is not None:
+            data = []
+            for subseries in series:
+                subseries_name = subseries.get('key')
+                # New name is "series_name.subseries_name" (if they exist)
+                data += self._es_rec(subseries['agg']['buckets'], '.'.join(filter(lambda x: x,
```
Comment (mahendra) [inline cabot/metricsapp/models/elastic.py]: You can do 
```python
'.'.join(filter(None, [series_name, subseries_name]))
```

Comment (mahendra) [inline cabot/metricsapp/models/elastic.py]: For readability, I would recommend assigning this to a separate variable)

File: cabot/metricsapp/api/elastic.py [hunk 18]
```diff
@@ -1,11 +1,54 @@
 from elasticsearch import Elasticsearch
+from enum import Enum
+
+
+class SupportedMetrics(Enum):
+    # Response includes the field "value"
+    METRICS_SINGLE = ['min', 'max', 'avg', 'value_count', 'sum', 'cardinality', 'moving_avg', 'derivative']
```
Comment (mahendra) [inline cabot/metricsapp/api/elastic.py]: Also, looking at your usage pattern, you should probably use `set()` instead of `list`

Also predefine
`METRICS_ALL = METRICS_SINGLE.union(METRICS_MULTIPLE)`

File: cabot/metricsapp/api/elastic.py [hunk 19]
```diff
@@ -1,11 +1,54 @@
...
     Create an elasticsearch-py client
     :param urls: comma-separated string of urls
+    :param timeout: timeout for queries to the client
     :return: a new elasticsearch-py client
     """
     urls = [url.strip() for url in urls.split(',')]
     return Elasticsearch(urls, timeout=timeout)
+
+
+def validate_query(query):
```
Comment (mahendra) [inline cabot/metricsapp/api/elastic.py]: Would this be a simpler / more readable function?

```python
def validate_query(query, msg_prefix='Elasticsearch query format error'):
    """
    Validate that an Elasticsearch query is in the format we want
    (all aggregations named 'agg', 'date_histogram' most internal
    aggregation, other metrics named the same thing as their metric
    type, e.g. max, min, avg...).
    :param query: the raw Elasticsearch query
    """

    # Loop through all the aggregations, stopping when we hit a date_histogram
    query = query.get('aggs', None)

    if query is None:
        return

    query = query.get('agg', None)

    if query is None:
        raise ValueError('{}: boo'.format(msg_prefix))

    if 'date_histogram' not in query:
        validate_query(query)
        return

    # Last case. We cannot go any deeper than this
    if 'agg' in query:
        raise ValueError('{}: date_histogram must be the innermost'
                         'aggregation (besides metrics)'.format(msg_prefix))

    # If we found a date_histogram the rest of the aggs should be metrics
    query = query.get('aggs', None)

    if query is None:
        raise ValueError('{}: query must include a metric'.format(msg_prefix))

    for metric, item in query.iteritems():
        if metric not in METRICS_SINGLE + METRICS_MULTIPLE:
            raise ValueError('{}: unsupported metric {}'.format(
                msg_prefix, metric))

        if metric not in item:
            raise ValueError('{}: metric name must be the same '
                             'as the metric type'.format(msg_prefix))
```

Issues I saw with the prefix function:
* We repeat the key fetch operations
```python
if not query.get('agg'):
   ...

query = query['agg']
```
We are doing the key get twice in two or three places. Not good for performance.
* using `iteritems()` is more efficient than iterating through keys and doing a get.

Comment (mahendra) [inline cabot/metricsapp/api/elastic.py]: `msg_prefix` can be in defs.

File: cabot/metricsapp/api/elastic.py (hunk 1 above)
Comment (mahendra) [inline cabot/metricsapp/api/elastic.py]: These can be constants in `defs.py`

File: cabot/metricsapp/api/elastic.py (hunk 2 above)
Comment (mahendra) [inline cabot/metricsapp/api/elastic.py]: the default value defined in `settings.py`

File: cabot/metricsapp/models/elastic.py [hunk 20]
```diff
@@ -37,3 +49,132 @@ def client(self):
...
+        try:
+            queries = json.loads(self.queries)
+        except ValueError as e:
+            logger.exception('Error loading Elasticsearch queries: {}'.format(self.queries))
+            parsed_data['error_code'] = type(e).__name__
+            parsed_data['error_message'] = str(e)
+            parsed_data['error'] = True
+            return parsed_data
+
+        for query in queries:
```
Comment (mahendra) [inline cabot/metricsapp/models/elastic.py]: Have you considered using ElasticSearch MultiSearch? http://elasticsearch-dsl.readthedocs.io/en/latest/search_dsl.html#multisearch

Comment (mahendra) [inline cabot/metricsapp/models/elastic.py]: It makes use of concurrent searches to elasticsearch - https://www.elastic.co/guide/en/elasticsearch/reference/current/search-multi-search.html

The concurrency setting for the query can be a parameter. We can configure it in settings.py - with a default of 4 (maybe)

File: cabot/metricsapp/api/elastic.py (hunk 1 above)
Comment (elainearbaugh) [inline cabot/metricsapp/api/elastic.py]: Oh yeah, constants in defs.py sounds much better!

The difference is just whether the metric is expected under value or values in the response. I guess we don't really have to distinguish between them though, we could just look for both "value" and "values" when we parse the json--I thought it'd be good to make sure the response metric is the type we expect, but seems unlikely we'd get weird responses since we're validating queries...

File: cabot/metricsapp/api/elastic.py (hunk 2 above)
Comment (elainearbaugh) [inline cabot/metricsapp/api/elastic.py]: Sure! what do you think the best default value is?

In my testing on stage I've had several timeouts with ES's default timeout of 10s (don't tell Libor), maybe 20-30? don't want to make it too high

File: cabot/metricsapp/tests/fixtures/metrics_series.yaml (hunk 8 above)
Comment (elainearbaugh) [inline cabot/metricsapp/tests/fixtures/metrics_series.yaml]: If there's no error, the series data wouldn't include error_message instead of having "error_message": None.

Doesn't belong in this PR though sorry!

File: cabot/metricsapp/models/elastic.py (hunk 5 above)
Comment (elainearbaugh) [inline cabot/metricsapp/models/elastic.py]: save() makes a lot more sends (y)

File: cabot/metricsapp/forms/elastic.py (hunk 10 above)
Comment (elainearbaugh) [inline cabot/metricsapp/forms/elastic.py]: Agreed that it shouldn't be here, but I'm thinking about moving it to model.clean() instead of model.save().

When you submit data in a form, if you raise a ValidationError in model.clean(), it'll be caught and you can show an error message on the form page. If an error is raised in model.save(), you'll go to a django error page (or just get the message "an unhandled exception occurred," which is way worse ux. 

If we make a Cabot API and we're creating/updating models not through forms, I think we'd want to call [full_clean](https://docs.djangoproject.com/en/1.11/ref/models/instances/#django.db.models.Model.full_clean)() on all models before saving to make sure there aren't any other issues with the input data (it also cleans fields and validates unique constraints, which I think we'd want). As long as we do that when we're saving we'll still be validating queries, so I think validation in model.clean() would work out for a cabot API.

Lmk if you see any issues with this!

File: cabot/metricsapp/models/elastic.py (hunk 13 above)
Comment (elainearbaugh) [inline cabot/metricsapp/models/elastic.py]: removed

File: cabot/metricsapp/models/elastic.py (hunk 16 above)
Comment (elainearbaugh) [inline cabot/metricsapp/models/elastic.py]: makes sense

File: cabot/metricsapp/tests/fixtures/es_multiple_metrics_terms.json (hunk 7 above)
Comment (elainearbaugh) [inline cabot/metricsapp/tests/fixtures/es_multiple_metrics_terms.json]: I would never!!

File: cabot/metricsapp/api/elastic.py (hunk 19 above)
Comment (elainearbaugh) [inline cabot/metricsapp/api/elastic.py]: Yeah this looks better!

File: cabot/metricsapp/models/elastic.py (hunk 20 above)
Comment (elainearbaugh) [inline cabot/metricsapp/models/elastic.py]: Sure, definitely better!

We could either configure max_concurrent_searches in settings, or we could do a per-ESSource implementation with a default (similar to how timeout is handled). The second way seems better to me but lmk if you think otherwise

File: cabot/metricsapp/api/elastic.py (hunk 1 above)
Comment (mahendra) [inline cabot/metricsapp/api/elastic.py]: you could do 
```python
b = x.get('value', x.get('values'))
if b is None:
    ...
```
or fancier ...
```python
look_for = set(['value', 'values'])
found = set(x.keys()).intersection(look_for).pop()
yaaay = x.get(found)
```

File: cabot/metricsapp/api/elastic.py (hunk 2 above)
Comment (mahendra) [inline cabot/metricsapp/api/elastic.py]: Let us start with 30 seconds?

File: cabot/metricsapp/forms/elastic.py (hunk 10 above)
Comment (mahendra) [inline cabot/metricsapp/forms/elastic.py]: Is model.clean() called on a model.save()?

Comment (mahendra) [inline cabot/metricsapp/forms/elastic.py]: I think moving to model.clean() is a better option. good call

File: cabot/metricsapp/models/elastic.py (hunk 20 above)
Comment (mahendra) [inline cabot/metricsapp/models/elastic.py]: per ES source implementation is a better call. We don't have to re-deploy for a change :-)

File: cabot/metricsapp/models/elastic.py (hunk 16 above)
Comment (mahendra) [inline cabot/metricsapp/models/elastic.py]: So, I have been thinking about ways to simplify this function. How does this look?

```python
import json
from collections import defaultdict


class Boo(object):
    def _parse_es_response(self, filename):
        earliest_point = 0
        output = []
        series = [json.load(open(filename))['aggregations']]

        data = defaultdict(list)
        for metric, (ts, value) in self._parse_series(series):
            if ts > earliest_point:
                data[metric].append((ts, value))

        for series, datapoints in data.iteritems():
            output.append(dict(series=series, datapoints=datapoints))

        return output

    def _parse_series(self, series, series_name=None):
        """
        Look through the json response recursively to go through all the
        aggregation buckets
        :param series: the aggregations part of the ES response
        :param series_name: the name of the series we're looking at
            ("key1.key2....metric_name")
        """

        # All aggregations should be named 'agg'
        for subseries in series:
            # If there are no more aggregations, we've reached the metric
            if subseries.get('agg') is None:
                # Note: I am sending the subseries inside
                results = self._get_metric_data(subseries, series_name)

            else:
                # New name is "series_name.subseries_name" (if they exist)
                key = subseries.get('key', 'metric')
                sub_series_name = '.'.join(filter(None, [series_name, key]))
                results = self._parse_series(subseries['agg']['buckets'],
                                             series_name=sub_series_name)

            for result in results:
                yield result

    def _get_metric_data(self, subseries, series_name):
        """
        Extract the metric name and values from a json series
        :param series: subset of the ES response that includes a list of
            metric values with date histogram keys
        :param series_name: the name of the series we're looking at
        """
        ts = subseries['key'] / 1000

        for metric_name, detail in subseries.iteritems():
            if type(detail) != dict:
                continue

            metric_name = '{}.{}'.format(series_name, metric_name)

            if 'value' in detail:
                yield (metric_name, (ts, detail['value']))

            elif 'values' in detail:
                for key, value in detail['values'].iteritems():
                    sub_metric_name = '{}.{}'.format(metric_name, key)
                    yield (sub_metric_name, (ts, value))

            else:
                raise NotImplementedError()


if __name__ == '__main__':
    import sys
    b = Boo()
    print(json.dumps(b._parse_es_response(sys.argv[1]), indent=1))
```

How does it work? Generators. The entire parsing operation uses generators. Advantage: we handle the data very lazily. More memory efficient. The parsing into output and filtering is done at the very end.

Also, do you think that the code looks simpler?

Comment (mahendra) [inline cabot/metricsapp/models/elastic.py]: There could be another way to parse this JSON - jsonpath

https://pypi.python.org/pypi/jsonpath-rw  (or https://pypi.python.org/pypi/jsonpath/)

Pick whichever is used by platform - if you think this fits the bill. (i think platform uses jsonpath-rw)

**Comment by mahendra** [comment]
@elainearbaugh Two questions on the PR
1. We are giving wildcard indices (eg: platform-prod-live-*) - we may have 100's of these. The query may span across all of these indices. Ideally, we should support templates - like metric-%Y-%m-%d and query just two days' worth of indices. (we should take the date duration as an ES setting). This could explain why some queries are taking 10+ seconds to run.
1. How are we planning on handling the "incomplete data" in the latest bucket? Did you decide on something?

For [1], you should talk to @gregsterin . He may have some ideas too.

**Comment by mahendra** [comment]
Please add test cases for percentiles etc.

File: cabot/metricsapp/models/elastic.py (hunk 16 above)
Comment (mahendra) [inline cabot/metricsapp/models/elastic.py]: Also, my code gets rid of the contentious Enum requirement ;-)

Comment (elainearbaugh) [inline cabot/metricsapp/models/elastic.py]: Code definitely looks simpler 👍 

I looked a little more at jsonpath-rw and I don't think we need to use it...most of the parsing we're doing is just looking for a key and doing something with it, so we don't really need a powerful json parsing library.

**Comment by elainearbaugh** [comment]
@mahendra: about wild card indices--I looked into it and Elasticsearch already supports date templating (see https://www.elastic.co/guide/en/elasticsearch/reference/current/date-math-index-names.html), so if you want data from today and yesterday you can set the index for the ESSource as `<metrics-whatever-{now/d}>,<metrics-whatever-{now/d-1d}>`. I just tried a query with this index value on stage and it's super fast now!! Will include more info about this in the help text for index. 

I think dropping the last bucket is the best idea, although it would mean we're ignoring a bunch of data for checks with longer group bys... I can implement that in this pr

I have a [test case for percentile](https://github.com/Affirm/cabot/blob/e24de7a9cdc4898965ccd6241e69bcf510ba6142/cabot/metricsapp/tests/test_elasticsearch.py), should I be testing something else also?

**Comment by elainearbaugh** [comment]
trigger build

**Comment by elainearbaugh** [comment]
@mahendra All changes are done except for Django test fixtures--I spent a while trying to make them work but ran into a few issues with django-polymorphic, which is used for StatusCheck polymorphism. Was thinking it might not be the effort it would take to fix it we're only creating 2 models--lmk if I should prioritize it

**Comment by mahendra** [comment]
reviewing now. Django-fixtures is not a priority. (please do open a separate ticket, if required, to track it)

File: cabot/metricsapp/models/elastic.py [hunk 21]
```diff
@@ -21,10 +32,23 @@ def __str__(self):
...
+                  '(<static_name{date_math_expr{date_format|time_zone}}>). For example, an index could be '
+                  '<metrics-{now/d}>,<metrics-{now/d-1d}>, resolving to "metrics-yyyy-mm-dd,metrics-yyyy-mm-dd, "'
+                  'for the past 2 days of metrics.',
+    )
     timeout = models.IntegerField(
-        default=60,
+        default=settings.ELASTICSEARCH_TIMEOUT,
         help_text='Timeout for queries to this index.'
     )
+    max_concurrent_searches = models.IntegerField(
```
Comment (mahendra) [inline cabot/metricsapp/models/elastic.py]: 👍

File: cabot/settings.py [hunk 22]
```diff
@@ -318,6 +318,13 @@
 
 # Image for service page
 SERVICE_IMAGE = os.environ.get('SERVICE_IMAGE', None)
+
+# Metricsapp settings
+# Timeout for Elasticsearch queries
+ELASTICSEARCH_TIMEOUT = os.environ.get('ELASTICSEARCH_TIMEOUT', 20)
```
Comment (mahendra) [inline cabot/settings.py]: `ELASTICSEARCH_TIMEOUT = int(...)` ?

File: cabot/settings.py [hunk 23]
```diff
@@ -318,6 +318,13 @@
...
+# msearch max_concurrent_searches
+ELASTICSEARCH_MAX_CONCURRENT_SEARCHES = os.environ.get('ELASTICSEARCH_MAX_CONCURRENT_SEARCHES', None)
```
Comment (mahendra) [inline cabot/settings.py]: `int( .. )` ?

File: cabot/settings.py (hunk 22 above)
Comment (elainearbaugh) [inline cabot/settings.py]: yes! thanks for catching

File: cabot/settings.py (hunk 23 above)
Comment (elainearbaugh) [inline cabot/settings.py]: this one's a little weird since it can be none (for the default concurrency setting)...I did it with code in settings.py :( but lmk if there's a better way

**Comment by mahendra** [comment]
reviewed latest changes. approval still holds.
